packages/cli/src/models/network/NetworkBaseController.js

```diff
@@ -122,6 +140,98 @@ export default class NetworkBaseController {
+                   `at the end of the contract.\n` +
+                   `  This does not alter the original storage, and should be a safe change.`)
+          break;
+        case 'pop':
+          log.warn(`- Variable '${originalVarDescription}' was removed from the end of contract ${original.contract}.\n`+
+                   `  Though this will not alter the positions in storage of other variables, if a new variable is added ` +
+                   `at the end of the contract, it will have the initial value of ${original.label} when upgrading. ` +
+                   `Avoid deleting existing variables to prevent this issue, and rename them to communicate that they are not to be used.`);
+          break;
+        case 'rename':
+          log.warn(`- Variable '${originalVarDescription}' in contract ${original.contract} was renamed to ${updated.label} in ${updatedVarSource}.\n` +
+                   `  Note that the new variable ${updated.label} will have the value of ${original.label} after upgrading. ` +
+                   `If this is not the desired behavior, add a new variable ${updated.label} at the end of your contract instead.`)
+          break;
+        case 'typechange':
+          log.warn(`- Variable '${original.label}' in contract ${original.contract} was changed from ${originalVarType.label} ` +
+                   `to ${updatedVarType.label} in ${updatedVarSource}.\n` + 
+                   `  If ${updatedVarType.label} is not compatible with ${originalVarType.label}, ` +
+                   `then ${updated.label} could be initialized with an invalid value after upgrading. Avoid changing types of existing ` +
+                   `variables to prevent this issue, and declare new ones at the end of your contract instead.`)
+          break;
+        case 'replace':
+          log.warn(`- Variable '${originalVarDescription}' in contract ${original.contract} was replaced with '${updatedVarDescription}' ` +
+                   `in ${updatedVarSource}.\nThis will cause ${updated.label} to be initialized with the value of ${original.label}. ` +
+                   `  If type ${updatedVarType.label} is not compatible with ${originalVarType.label}, then ${updated.label} could be `+
+                   `initialized with an invalid value after upgrading. Avoid changing types of existing ` +
+                   `variables to prevent this issue, and declare new ones at the end of your contract instead.`)
+          break;
+        default:
+          log.error(`Unexpected layout changeset: ${action}`)
+      }
+    });
+  }
+
+  _srcToLineNumber(sourceCode, srcFragment) {
+    if (!sourceCode || !srcFragment) return null;
+    const [begin] = srcFragment.split(':', 1);
+    return sourceCode.substr(0, begin).split('\n').length
+  }
+
+  _tryGetSourceFragment(sourceCode, src) {
+    if (!src || !sourceCode) return null;
+    const [begin, count] = src.split(':');
+    return sourceCode.substr(begin, count);
+  }
+
```

Comment:
I'd move all this logic outside the controller

Reply:
Absolutely!